src/main/java/org/broadinstitute/hellbender/tools/sv/SVCallRecord.java

```diff
@@ -1,279 +1,172 @@
                         final int length,
                         final List<String> algorithms,
-                        final List<Genotype> genotypes) {
-        Utils.nonNull(startContig);
-        Utils.nonNull(endContig);
+                        final List<Allele> alleles,
+                        final List<Genotype> genotypes,
+                        final Map<String,Object> attributes) {
+        Utils.nonNull(id);
+        Utils.nonNull(contigA);
+        Utils.nonNull(contigB);
         Utils.nonNull(type);
         Utils.nonNull(algorithms);
+        Utils.nonNull(alleles);
         Utils.nonNull(genotypes);
-        Utils.nonEmpty(algorithms);
-        Utils.nonEmpty(genotypes);
-        Utils.containsNoNull(algorithms, "Encountered null algorithm");
-        Utils.containsNoNull(genotypes, "Encountered null genotype");
-        this.startContig = startContig;
-        this.start = start;
-        this.startStrand = startStrand;
-        this.endContig = endContig;
-        this.end = end;
-        this.endStrand = endStrand;
+        Utils.nonNull(attributes);
+        this.id = id;
+        this.contigA = contigA;
+        this.positionA = positionA;
+        this.strandA = strandA;
+        this.contigB = contigB;
+        this.positionB = positionB;
+        this.strandB = strandB;
         this.type = type;
         this.length = length;
-        this.algorithms = algorithms;
-        this.genotypes = genotypes;
-        this.samples = genotypes.stream()
-                .filter(g -> g.hasExtendedAttribute(GATKSVVCFConstants.COPY_NUMBER_FORMAT)) //skip no-calls that don't have copy number (i.e. not dupe no-calls)
-                .map(Genotype::getSampleName)
-                .collect(Collectors.toCollection(LinkedHashSet::new));
+        this.algorithms = Collections.unmodifiableList(algorithms);
+        this.alleles = Collections.unmodifiableList(alleles);
+        this.altAlleles = alleles.stream().filter(allele -> !allele.isNoCall() && !allele.isReference()).collect(Collectors.toList());
+        final List<Allele> refAllelesList = alleles.stream().filter(allele -> !allele.isNoCall() && allele.isReference()).collect(Collectors.toList());
+        Utils.validate(refAllelesList.size() <= 1, "Encountered multiple reference alleles");
+        this.refAllele = refAllelesList.isEmpty() ? null : refAllelesList.get(0);
+        this.genotypes = GenotypesContext.copy(genotypes).immutable();
+        this.attributes = Collections.unmodifiableMap(attributes);
+    }
+
+    public SVCallRecord(final String id,
+                        final String contigA,
+                        final int positionA,
+                        final boolean strandA,
+                        final String contigB,
+                        final int positionB,
+                        final boolean strandB,
+                        final StructuralVariantType type,
+                        final int length,
```

Comment:
I've made length and the strands into `Integer` and `Boolean` types, respectively, so that they can take on null values when appropriate (e.g. length for BNDs or strands for CNVs). This required quite a bit of refactoring and fixing of tests but I seem to have it working. There may be some performance hit from boxing/unboxing these values everywhere but I do agree that passing around -1 everywhere (and having non-sensical strand values) is problematic in some ways.

It is a little tricky to have constructors that allow for length/strand parameters that are potentially inconsistent, so I have checks to ensure the values are consistent and also sanitize attributes that could contain redundant information.